onnxruntime/core/providers/cuda/tensor/transpose.h

```diff
@@ -14,12 +14,15 @@ namespace cuda {
 
 class Transpose final : public CudaKernel, public TransposeBase {
  public:
-  Transpose(const OpKernelInfo& info) : CudaKernel(info), TransposeBase(info) {}
+  Transpose(const OpKernelInfo& info) : CudaKernel(info), TransposeBase(info) {
+  }
 
   Status ComputeInternal(OpKernelContext* context) const override;
 
-  static Status DoTranspose(const Transpose& transpose_kernel,
-                            const std::vector<size_t>& permutations, const Tensor& input, Tensor& output);
+  // `input_shape_override` overrides the shape of `input` for compute purposes.
+  static Status DoTranspose(const cublasHandle_t cublas_handle,
```

Comment:
Change the interface a bit so that a full Transpose kernel instance is not required by the static method that actually does the Transpose operation. All call sites may not be able to construct a full Transpose CUDA kernel. So, limiting the interface to just the "tool" needed to perform the transpose operation in a stand-alone fashion. The only "tool" required in the CuBlas handle.

Reply:
EDIT: It looks like this method was used in CUDA scan and in a training kernel. Rather than change the interface and make changes in all the places it is called to hold a non-const CUDA EP (to get the cublas handle to pass in here), added an overload of DoTranspose that takes just the CuBlas handle and make use of the new method in Einsum and keeping existing code in CUDA scan and the training kernel as is